crates/guest-rust/rt/src/async_support/cabi.rs

```diff
@@ -0,0 +1,105 @@
+//! knowledge of 0.B.0 meaning that when 0.B.0 decides to block it won't know
+//! where to register its `waitable` within a `waitable-set`.
+//!
+//! To solve this problem the long-term intention is that something will live
+//! in `wasi-libc` itself, but in the meantime it's living "somewhere" within
+//! `wit-bindgen 0.*.0`. Specifically all `wit-bindgen` versions will all
+//! reference, via C linkage, a single function which is used to manipulate a
+//! single pointer in linear memory. This pointer is a `wasip3_task` structure
+//! which has all the various fields to use it.
+//!
+//! The `wasip3_task_set` symbol is itself defined in C inside of the
+//! `src/wit_bindgen_cabi.c` file at this time, specifically because it's
+//! annotated with `__weak__` meaning that any definition of it suffices. This
+//! isn't possible to define in stable Rust (specifically `__weak__`).
+//!
+//! Once `wasip3_task_set` is defined everything then operates via indirection,
+//! aka based off the returned pointer. The intention is that exported functions
+//! will set this (it's sort of like an executor) and then imported functions
+//! will all use this as the source of registering waitables. In the end that
+//! means that it's possible to share types with `std::future::Future` that
+//! are backed at the ABI level with this "channel".
+//!
+//! In the future it's hoped that this can move into `wasi-libc` itself, or if
+//! `wasi-libc` provides something else that would be prioritized over this.
+//! For now this is basically an affordance that we're going to be frequently
+//! releaseing new major versions of `wit-bindgen` and we don't want to force
+//! applications to all be using the exact same version of the bindings
+//! generator and async bindings.
+//!
+//! Additionally for now this file is serving as documentation of this
+//! interface.
+
+use core::ffi::c_void;
+
+extern "C" {
+    /// Sets the global task pointer to `ptr` provided. Returns the previous
+    /// value.
+    ///
+    /// This function acts as both a dual getter and a setter. To get the
+    /// current task pointer a dummy `ptr` can be provided (e.g. NULL) and then
+    /// it's passed back when you're done working with it. When setting the
+    /// current task pointer it's recommended to call this and then call it
+    /// again with the previous value when the tasks's work is done.
+    ///
+    /// For executors they need to ensure that the `ptr` passed in lives for
+    /// the entire lifetime of the component model task.
+    pub fn wasip3_task_set(ptr: *mut wasip3_task) -> *mut wasip3_task;
+}
+
+/// The first version of `wasip3_task` which implies the existence of the
+/// fields `ptr`, `waitable_register`, and `waitable_unregister`.
+pub const WASIP3_TASK_V1: u32 = 1;
+
+/// Indirect "vtable" used to connect imported functions and exported tasks.
+/// Executors (e.g. exported functions) define and manage this while imports
+/// use it.
```

Comment:
Note to self. Why am I making comments when I can't even visual what is meant by exporting a task?